bam_readcount_helper.py takes in flags from bam_readcount.cwl

johnegarza opened the pull request:
Modified bam_readcount_helper.py to take in -q and -b flags from bam_readcount.cwl. Part of issue #214 in cancer-genomics-workflow

johnegarza committed:
made -q and -b flags user definable

tmooney commented on bam_readcount_helper.py:
```diff
@@ -16,8 +16,14 @@ def generate_region_list(hash):
+        bam_readcount_cmd.append('20')
+    else:
+        bam_readcount_cmd.append(str(base_qual))
+    if map_qual != -1: #default value in bam_readcount.cwl
```
It might be nicer to have these be optional parameters instead of passing a special value.  (Maybe we're at the point of wanting this helper to use `argparse`?)

Until we get better versioning of Docker images in our pipelines, we may need to revise existing processing profiles that include this workflow step if we break backwards compatibility with the current version of the helper.

johnegarza committed:
updating default value logic

johnegarza commented:
In response to the previous comment from @tmooney I changed the logic for adding the -b (minimum base quality) flag value into the call to bam-readcount. Since the script originally hardcoded the value as 20, I went back to bam_readcount.cwl in cancer-genomics-workflow and made its default value 20 there (see PR #320). This means the script no longer has to check that value, and can simply add it in into the call (line 21 of the current commit). However, the -q (minimum mapping quality) flag did not have a default, and it is now completely optional in bam_readcount.cwl, so I updated the script to check the length of `sys.argv` and set variables accordingly. This could be simplified in the same way as the -b flag; bamreadcount.cpp defaults the value of -q to 0, so bam_readcount.cwl could be updated to reflect this, but this means any changes to the default value in bamreadcount.cpp would have no effect unless bad_readcount.cwl is updated alongside it.

So: is the current logic that checks `sys.argv` to set variables acceptable, or should -q/minimum mapping quality just be given a default value of 0 in bam_readcount.cwl

susannasiebert commented on bam_readcount_helper.py:
```diff
@@ -16,8 +16,11 @@ def generate_region_list(hash):
-    bam_readcount_cmd = ['/usr/bin/bam-readcount', '-f', ref_fasta, '-l', region_list, '-w', '0', '-b', '20']
+def filter_sites_in_hash(region_list, bam_file, ref_fasta, sample, output_dir, insertion_centric, map_qual, base_qual):
+    bam_readcount_cmd = ['/usr/bin/bam-readcount', '-f', ref_fasta, '-l', region_list, '-w', '0', '-b']
+    bam_readcount_cmd.append(str(base_qual))
```
Small nit-pick but you can just add `str(base_qual)` to the array on line 20 directly.

susannasiebert reviewed (changes requested):
If we are concerned about backwards compatibility, we could change this to:
1) pass `min_mapping_qual` and `min_base_qual` to the script as the last arguments instead of in the middle
2) set default values for them when they are not passed in

johnegarza committed:
change argument ordering and use default values

susannasiebert commented on bam_readcount_helper.py:
```diff
@@ -35,11 +32,7 @@ def filter_sites_in_hash(region_list, bam_file, ref_fasta, sample, output_dir, i
-    (script_name, min_base_qual, vcf_filename, sample, ref_fasta, bam_file, output_dir)= sys.argv
-elif len(sys.argv) == 8:
-    (script_name, min_mapping_qual, min_base_qual, vcf_filename, sample, ref_fasta, bam_file, output_dir)= sys.argv
+(script_name, vcf_filename, sample, ref_fasta, bam_file, output_dir, min_mapping_qual, min_base_qual)= sys.argv
```
I think I wasn't quite clear how this should be changed. You basically have to handle three cases:
1) `len(sys.argv) == 6` i.e, backwards compatibility where neither mapping quality nor base quality values are passed along
2) `len(sys.argv) == 7` i.e., when only min_base_qual is passed along
3) `len(sys.argv) == 8` i.e, both min_base_qual and min_mapping_qual are provided. Additionally, to be consistent with case (2) above, you'll want to provide the min_base_qual before the min_mapping_qual.
For cases (1) and (2) you will also set those variables that are not passed along to a default.

johnegarza committed:
changes to fully support backward compatibility

chrisamiller commented:
seems fine from a cursory look - +1
